[`Ernie 4.5`] Ernie VL models

vasqu opened the pull request:
Continuation of #39228 for the VL models

Current inference script for testing (torch 2.9.1):
```python
import requests
from PIL import Image

from transformers import AutoConfig, AutoModelForImageTextToText, AutoProcessor


model_path = "/raid/anton/code/forks/transformers/AntonV/ErnieVL"
processor = AutoProcessor.from_pretrained(model_path)

config = AutoConfig.from_pretrained(model_path)
model = AutoModelForImageTextToText.from_pretrained(
    "baidu/ERNIE-4.5-VL-28B-A3B-PT",
    config=config,
    device_map="auto",
    dtype="auto",
)

messages = [
    {
        "role": "user",
        "content": [
            {"type": "text", "text": "Only use English during your responses and describe the following image."},
            {"type": "image"},
        ]
    },
]
text = processor.apply_chat_template(messages, tokenize=False, add_generation_prompt=True)
image = Image.open(requests.get("https://paddlenlp.bj.bcebos.com/datasets/paddlemix/demo_images/example1.jpg", stream=True).raw)
inputs = processor(text=[text], images=[image], return_tensors="pt").to(model.device)

generated_ids = model.generate(
    **inputs,
    max_new_tokens=64,
    do_sample=False,
)
print(processor.decode(generated_ids[0][len(inputs['input_ids'][0]):]))
```
Output: 
`The image features a scenic view of rolling mountains under a clear sky. In the foreground, there is a person sitting on a hillside, facing away from the camera. The person is wrapped in a colorful, striped blanket. The landscape is dotted with vibrant pink flowers, adding a pop of color to the`

Left TODOs:
- [x] TP
  - [ ] EP to add later after its been fixed

vasqu committed:
more attention cleanup

vasqu committed:
llama like text attention

vasqu committed:
generates different text but cos and sin tensors are always close - 1e-8

vasqu committed:
another round of rope fixups

vasqu committed:
yea, gonna check tomorrow cant cheat w freqs for whatever reason

vasqu committed:
NOTE: last time where comp with old rope

vasqu committed:
rope cleanup

vasqu committed:
more rope

vasqu committed:
somewhat clean 3d rope with attn - sin / cos has very small diffs to original formula (torch.allclose always True) leading to slightly different generations

vasqu committed:
new rope type

vasqu committed:
style

vasqu committed:
attempt at moe, gonna need a deeper look

vasqu committed:
cleanup gate

vasqu committed:
more cleaning

vasqu committed:
NOTE remove attempt at moe for now

vasqu committed:
another round of cleanups

vasqu committed:
whoops

vasqu committed:
we back boys, reattempting moe start

vasqu committed:
moe should be done with this

vasqu committed:
cleanup

vasqu committed:
more cleanup

vasqu committed:
nits

vasqu committed:
add conversion and adjust code accordingly

vasqu committed:
fix

vasqu committed:
Merge branch 'main' into ernie_vl

vasqu committed:
make moe copyable as far as we can

vasqu committed:
cleanup conversion a bit, next config

vasqu committed:
cleanup config part1

vasqu committed:
small removal of unused things

vasqu committed:
config conversion, rope type doesnt get loaded tho...

vasqu committed:
fix rope

vasqu committed:
last hardcoded values

vasqu committed:
remove unnecessary class

vasqu committed:
starting to make copies available for vision, vision rope refactor tomorrow

vasqu committed:
vl rope changes

vasqu committed:
simplify variable resolution resampler

vasqu committed:
nit

vasqu committed:
conversion update

vasqu committed:
more conversions, standardization, and big dtype fix!

vasqu committed:
remove some docs (tmp), focus on code for me

vasqu committed:
oops

vasqu committed:
nit

vasqu committed:
fixup embeddings, add todos

vasqu committed:
more cleanup

vasqu committed:
more cleanup, next caching changes

vasqu committed:
revert fp16, internally discussed weights are supposed to be bf16

vasqu committed:
fix rope (a bit), prepare cache logic changes

vasqu committed:
more prep for cache

vasqu committed:
cache class is used, fixup some flags

vasqu committed:
modular refactor

vasqu committed:
partially docstrings, docs, etc

vasqu committed:
cleaner order

vasqu committed:
nit

vasqu committed:
fix config

vasqu committed:
remove old artefacts/todos

vasqu committed:
Merge branch 'main' into ernie_vl

vasqu committed:
sync with remote and add some todos for orientation

vasqu committed:
remove img process dep on modeling code

vasqu committed:
image processor with a few diffs highlighted to copy from maybe

vasqu committed:
fast img processor version

vasqu committed:
modular image processors

vasqu committed:
convert tokenizer to have dedicated video placeholder token

vasqu committed:
before i forget

vasqu committed:
Merge branch 'main' into ernie_vl

vasqu committed:
a modular bug :/

vasqu committed:
more processor things, some modular adjustments

vasqu committed:
remove dependency on token type ids

vasqu committed:
position ids ala qwen vl and modular is bugging

vasqu committed:
fixup some inheritances + nits

vasqu committed:
token type ids

vasqu committed:
moe loss, docs, simplify pos ids

vasqu committed:
align some feature getters

vasqu committed:
docs

vasqu committed:
rename conv -> merge aka our naming convention

vasqu committed:
style

vasqu committed:
fixup tokenizer class in auto

vasqu committed:
no more nn sequential

vasqu committed:
fix chat template, fix tokenizer conversion, modular bug

vasqu committed:
remove this

vasqu committed:
remove old deps (from the remote processor)

vasqu committed:
Merge branch 'main' into ernie_vl

vasqu committed:
whoops

vasqu committed:
argh

vasqu committed:
todo, restarting progress tomorrow

vasqu committed:
fast image processor changes output, keeping slow for now

vasqu committed:
NOTE rm debugging code on processor conversion

vasqu committed:
first complete conversion script version, todo on whether to use fast processor

vasqu committed:
config docs

vasqu committed:
image processor tests, only kept to images as videos need different resolutions

vasqu committed:
Merge branch 'main' into ernie_vl

vasqu committed:
processor tests

vasqu committed:
first ish version for video processor, very much WIP tho

vasqu committed:
Merge branch 'main' into ernie_vl

vasqu committed:
sync with main and all the changes that happened, fix ernie moe bug in dtype casting

vasqu committed:
mini style fix

vasqu committed:
Merge branch 'main' into ernie_vl

vasqu committed:
vid processor is properly separated now

vasqu committed:
make vid processor its own thing

vasqu committed:
style

vasqu committed:
video processing and cleanups, img processing done, processing needs one TODO, vid processing needs tests

vasqu committed:
readd vid patch fn

vasqu committed:
make 4D RoPE possible if manually passed

vasqu committed:
simplify the msg on packing, allow external prep but not internal one

vasqu committed:
nit

vasqu committed:
revert general changes video utils, make it specific to ernie, fixup tests

vasqu committed:
vid to auto

vasqu committed:
left to check: pos ids (rope) + token type ids

vasqu committed:
move token type ids to processor, fix processor to ernie logic

TODOs: tests, tests, tests

vasqu committed:
processor fixes, conversion todo for fast img processor

TODOs: tests for vid processor and modeling

vasqu committed:
fix

vasqu committed:
video processor tests, torch compile does not work due to PIL drawing being needed

vasqu committed:
fix config consistency

vasqu committed:
style

vasqu committed:
wip tests

vasqu committed:
fix most tests, 2 failing ones remain

vasqu committed:
fix last tests

vasqu committed:
check

vasqu committed:
docs consistency

vasqu committed:
fix conversion script, more docs

vasqu committed:
optional drawing on frames, style

vasqu committed:
add error on compile x draw on frames

vasqu committed:
fix

vasqu committed:
fix

vasqu committed:
change font loading to hub dep with default font

vasqu committed:
fix config try 2

vasqu committed:
fix diff resolution, tests (not fast processor, a100)

vasqu committed:
fix test

vasqu committed:
style

vasqu committed:
Merge branch 'main' into ernie_vl

vasqu committed:
torch 2.9 (fa2 untested, video from 2.6)

zucchini-nlp commented on src/transformers/models/auto/modeling_auto.py:
```diff
@@ -980,6 +981,7 @@ class _BaseModelWithGenerate(PreTrainedModel, GenerationMixin):
         ("blip", "BlipForConditionalGeneration"),
         ("blip-2", "Blip2ForConditionalGeneration"),
         ("chameleon", "ChameleonForConditionalGeneration"),
+        ("ernie4_5_vl", "Ernie4_5_VLForConditionalGeneration"),
```
let's delete from here, we'll get rid of `VISION_2_SEQ` some time in the future. Hopefully soon

vasqu replied:
Gotcha, removing it

zucchini-nlp commented on src/transformers/models/ernie4_5_vl/modular_ernie4_5_vl.py:
```diff
@@ -0,0 +1,1495 @@
+
+
+class Ernie4_5_VLVisionBlock(Qwen2_5_VLVisionBlock):
+    def __init__(self, config, attn_implementation: str = "sdpa") -> None:
```
not related to this PR: `attn_implementation` doesn't seem to be used, so we should delete from inputs args in qwenVL

zucchini-nlp commented on src/transformers/models/ernie4_5_vl/modular_ernie4_5_vl.py:
```diff
@@ -0,0 +1,1495 @@
+    def forward(self, x, grid_thw):
+        # image spatial
+        x = x.reshape([-1, x.shape[-1] * (self.spatial_merge_size**2)])
+        x = self.spatial_linear(x.to(self.mlp.weight.dtype))
```
hidden states are casted to a MLP's dtype and then fed into another layer. I think dtypes could potentially mismatch ?

zucchini-nlp commented on src/transformers/models/ernie4_5_vl/modular_ernie4_5_vl.py:
```diff
@@ -0,0 +1,1495 @@
+        prefill_noncompiled_stage = not is_torchdynamo_compiling() and (
+            (cache_position is not None and cache_position[0] == 0)
+            or (past_key_values is None or past_key_values.get_seq_length() == 0)
+        )
```
oh, it's something I've been trying to chnage with https://github.com/huggingface/transformers/pull/42088. Current logic is definitely not bug-free. Let's add a TODO for us, moving it all to generation-side would be the best option

And then let users handle position ids themselves if they're running `forward` with manual generation loop

vasqu replied:
Adding a TODO and referencing the PR for now

zucchini-nlp commented on src/transformers/models/ernie4_5_vl/modular_ernie4_5_vl.py:
```diff
@@ -0,0 +1,1495 @@
+
+    @property
+    def visual(self):
+        return self.model.vision_tower
```
this has resulted to be a bad pattern which I am planning to delete, and break BC. Initially it was added for some models only because I rearrange completely the model layout. For new models let's not add it

Instead I plan to have a general utility to get sub-models like `self.get_vision_module()` etc

molbap replied:
oh that's cool, came across this pattern many times. looking forward to the PR!

vasqu replied:
I initially removed the property, but `test_sdpa_can_dispatch_composite_models` breaks - just FYI. Reverted for now, ig you will change this in a different PR

zucchini-nlp commented on src/transformers/models/ernie4_5_vl/processing_ernie4_5_vl.py:
```diff
@@ -0,0 +1,290 @@
+    image_processor_class = "AutoImageProcessor"
+    video_processor_class = "AutoVideoProcessor"
+
+    tokenizer_class = (None, "LlamaTokenizerFast")
```
maybe `AutoTokenizer`, especially after merging https://github.com/huggingface/transformers/pull/41633

vasqu replied:
I've had enough pain there. It's a problem where a weird discrepancy between llama slow and fast tokenizers came apparent - especially when using special tokens. I don't want to introduce possibly faulty tokenizers and other models also often don't have a slow tokenizer defined

Edit: maybe past ernie PRs can show you some past lol

vasqu replied:
Actually that makes a lot of sense, I kinda forgot about that PR; will leave this open until I merge with main and resolve things step by step

yonigozlan replied:
Just seeing this sorry. No need to have any of these anymore (they are not used anyway). It will just use what in the auto classes Also no need to have the `attributes` attribute, they are now parsed from the init signature.

vasqu replied:
Thx, I'm a bit behind on main before your PR so will keep this in mind when I merge main and resolve the last things bit by bit. Was avoiding it due to the weight converter which will need another careful look :D

zucchini-nlp commented on src/transformers/models/ernie4_5_vl/video_processing_ernie4_5_vl.py:
```diff
@@ -0,0 +1,469 @@
+        # by the user, it gets its default value from the instance, or is set to None.
+        for kwarg_name in self.valid_kwargs.__annotations__:
+            if "fps" in kwarg_name:  # we ignore fps from self
+                kwargs.setdefault(kwarg_name, None)
```
why is that ignored?

vasqu replied:
It's been a while but in essence, these models do not fps at all and we should always provide the metadata but in case we don't it should use our own `self.fps`. However, when I do not explicitly do not set this here, I will clash with fps because it thinks I provided fps which is not the case (inferred from self) --> I handle that logic myself.

vasqu replied:
Second attempt as to why this is necessary: Ernie does not use a default fps at all and only samples based on the number of frames. However, this requires specific metadata (e.g. fps iirc to get proper time stamps) to be available along the video, which is not guaranteed.

We do not want to use fps as requested kwarg, instead it will be used as fallback for the metadata if the fps is not available; it also invokes a warning in those cases in `_prepare_input_videos`.

This requires custom logic where we do not want to default to using fps except when a user explicitly requests it. However, the loop overrides the fps kwarg; we don't want that! The user is allowed to but he shouldn't do it.

Tl;dr: Fps is has custom logic and is intended to be used as fallback for metadata, not as actual requested fps by default.

molbap commented on src/transformers/models/ernie4_5_moe/modeling_ernie4_5_moe.py:
```diff
@@ -316,11 +316,12 @@ def forward(self, hidden_states):
+        intermediate_size = config.moe_intermediate_size if intermediate_size is None else intermediate_size
         for _ in range(self.num_experts):
-            self.append(Ernie4_5_MoeMLP(config, config.moe_intermediate_size))
+            self.append(Ernie4_5_MoeMLP(config, intermediate_size))
```
bit weird to have to instantiate a modulelist with more than the config, is it a varying arg? (i.e. not entirely config-derivable)

molbap commented on src/transformers/models/ernie4_5_vl/configuration_ernie4_5_vl.py:
```diff
@@ -0,0 +1,314 @@
+
+    model_type = "ernie4_5_vl_text"
+    base_config_key = "text_config"
+    attribute_map = {"num_experts": "moe_num_experts", "num_experts_per_tok": "moe_k"}
```
I think https://github.com/huggingface/transformers/pull/41580 will come into play here :eyes:

molbap commented on src/transformers/models/ernie4_5_vl/convert_ernie4_5_vl_to_hf.py:
```diff
@@ -0,0 +1,465 @@
+            if "moe_statics" in key:
+                suffix = "moe_statics.e_score_correction_bias"
+                converted_key = key.removesuffix(suffix)
+                # splitting param (2, ...) to 2 * (1, ...)
```
Instead of having several cases here, maybe would be more readable to have an `ORIGINAL_TO_CONVERTED_KEY_MAPPING` now that all mappings are established? (also should be easier to then port to the new model loading interface)

molbap commented on src/transformers/models/ernie4_5_vl/convert_ernie4_5_vl_to_hf.py:
```diff
@@ -0,0 +1,465 @@
+    processor = Ernie4_5_VLProcessor(
+        # Intentionally use the slow image processor as the fast processor
+        # creates too much fluctuation affecting the model output
+        # image_processor=Ernie4_5_VLImageProcessorFast(),
```
erg that's too bad. If you have a notebook/script with the differences it would be interesting, I struggled a lot recently on [DeepseekOCR](https://github.com/huggingface/transformers/pull/41797) but now seems to work-ish

vasqu replied:
Hmm, the issue is that the image processor tests do pass - so the initial differences are not that big. Ig the size of the model comes into play + smaller (but necessary) changes on RoPE + the fp16 stuff we discussed.

No real script, you would need to manually convert with the fast processor instead and observe the different output, e.g. using the small script in the PR description

molbap commented on src/transformers/models/ernie4_5_vl/modular_ernie4_5_vl.py:
```diff
@@ -0,0 +1,1495 @@
+        else:
+            final_hidden_states, router_logits = self.text_moe(hidden_states)
+            final_hidden_states = final_hidden_states.reshape(batch_size, sequence_length, hidden_dim)
+            router_logits = router_logits.reshape(-1, self.num_experts)
```
this is interesting. Any way to always compute the `text_moe` with a mask whose masking operation is a no-op in text-only mode? to reduce a bit LoC here

molbap replied:
sure, it's mostly prettifying, does not change anything at runtime and clear enough as it is. SGTM

molbap commented on src/transformers/models/ernie4_5_vl/modular_ernie4_5_vl.py:
```diff
@@ -0,0 +1,1495 @@
+        )
+        cu_seqlens = F.pad(cu_seqlens, (1, 0), value=0)
+
+        for blk in self.blocks:
```
nit, full words (though it follows the qwen2.5VL style (which would benefit from `layer_types` by the way, separate issue))

molbap commented on src/transformers/models/ernie4_5_vl/modular_ernie4_5_vl.py:
```diff
@@ -0,0 +1,1495 @@
+        video_embeds = torch.split(video_embeds, split_sizes)
+        return video_embeds
+
+    def get_image_features(self, pixel_values: torch.FloatTensor, image_grid_thw: Optional[torch.LongTensor] = None):
```
formally, is this and `get_video_features` the same?

vasqu replied:
Mostly, except for its split sizes which comes from the resampler model.

molbap commented on src/transformers/models/ernie4_5_vl/modular_ernie4_5_vl.py:
```diff
@@ -0,0 +1,1495 @@
+
+@auto_docstring
+class Ernie4_5_VLImageProcessorFast(Qwen2VLImageProcessorFast):
+    size = {"shortest_edge": 56 * 56, "longest_edge": 28 * 28 * 6177}
```
I lost track, so now defaults should be held here? Q @zucchini-nlp too

molbap replied:
no nothing changed haha, it's my brain that struggled there one min

molbap reviewed (commented):
Thanks for the huge work! Left a review after reading that of @zucchini-nlp

molbap commented on src/transformers/models/ernie4_5_vl/modular_ernie4_5_vl.py:
```diff
@@ -0,0 +1,1495 @@
+        freq_hw = torch.stack([freq_h, freq_w], dim=-1).flatten(-2)
+        freq_hwt = torch.cat([freq_hw, freq_t], dim=-1)
+        return freq_hwt.repeat_interleave(2, dim=-1)
+
```
IMO: the _current_ mrope implems are hard to read and would benefit some standardisation, but it does not befall to this PR to solve it indeed.
I remember what we discussed about the additional rotations and after a back-of-the-envelope doodle (you have certainly explored more) I don't see an easy refactoring here and now. Agree that overriding the init can be misleading, maybe possible with a big big disclaimer?

Definitely a big TODO overall for later

vasqu replied:
Can move this to a default init instead for now - adding a fat disclaimer. Are you on boat with this @zucchini-nlp

zucchini-nlp replied:
> IMO: the current mrope implems are hard to read and would benefit some standardisation, but it does not befall to this PR to solve it indeed.
+ 1, totally agreed and it was moved to the backlog a lot of times by me 😓 

We can be explciit and use `ernie_3d` as rope-type but I think it doesn't belong in general utilities. Instead all the RoPE code will be in the model code and we won't support dynamic RoPE types.

vasqu commented on src/transformers/models/ernie4_5_vl/modular_ernie4_5_vl.py:
```diff
@@ -0,0 +1,1495 @@
+        """
+        output_router_logits = (
+            output_router_logits if output_router_logits is not None else self.config.text_config.output_router_logits
+        )
```
Yes, that's missing in `check_model_inputs()`. It's a bigger issue and covers more models!

vasqu commented on src/transformers/models/ernie4_5_vl/modular_ernie4_5_vl.py:
```diff
@@ -0,0 +1,1495 @@
+
+    def _preprocess(
+        self,
+        images: Union[ImageInput, VideoInput],
```
Yup, seems to be deprecated now. However, note that qwen vls still have the wrong signature too :skull: changing it here

vasqu commented on src/transformers/models/ernie4_5_vl/modular_ernie4_5_vl.py:
```diff
@@ -0,0 +1,1495 @@
+                - `"channels_last"` or `ChannelDimension.LAST`: image in (height, width, num_channels) format.
+                - `"none"` or `ChannelDimension.NONE`: image in (height, width) format.   - `"none"` or `ChannelDimension.NONE`: image in (height, width) format.
+        """
+        images = make_list_of_images(images)
```
Might need to be updated in qwen vl too! Ig you mean `images = self.fetch_images(images)` before?

vasqu commented on src/transformers/models/ernie4_5_vl/processing_ernie4_5_vl.py:
```diff
@@ -0,0 +1,290 @@
+            if getattr(tokenizer, "image_token_id", None)
+            else tokenizer.convert_tokens_to_ids(self.image_token)
+        )
+        self.image_end_token_id = (
```
This is a bit more complicated. I'm not sure what you had in mind initially but the tokenizer does not support special tokens as called e.g. `image_token`; these are only reserved here https://github.com/huggingface/transformers/blob/240d19f4a38087ac4fc38c396f57a9c63a45c187/src/transformers/tokenization_utils_base.py#L842-L851

I initially tried to override them in the first ernie models but was stopped by Arthur. Currently, it is impossible IMO except you have a solution I could do here. It's important that these tokens still are special tokens so they somehow have to be in that list. Atm, they are under `additional_special_tokens`

vasqu replied:
Oh wow, this is neat TIL

vasqu commented on src/transformers/models/ernie4_5_vl/processing_ernie4_5_vl.py:
```diff
@@ -0,0 +1,290 @@
+                self.video_start_token_id,
+                self.video_end_token_id,
+            ]:
+                mm_token_type_ids[array_ids == token_id] = 1
```
Should I change this then? This was initially just to differentiate between img/vid and text tokens: There are different MoEs for each modality there. 

Also note that while start/end tokens are included they are special tokens while we still treat them as img/vid tokens.

vasqu committed:
raushan's review (part 1)

vasqu committed:
Update docs/source/en/model_doc/ernie4_5_vl.md

Co-authored-by: Pablo Montalvo <39954772+molbap@users.noreply.github.com>

vasqu commented on docs/source/en/model_doc/ernie4_5_vl.md:
```diff
@@ -0,0 +1,97 @@
+
+## Overview
+
+The ernie4_5_vl model was proposed in [<INSERT PAPER NAME HERE>](<INSERT PAPER LINK HERE>) by <INSERT AUTHORS HERE>.
```
Yea keeping one comment open here, I have a TODO note in the PR description :+1:

jzhang533 replied:
https://arxiv.org/abs/2510.14528 is the technical report for PaddleOCR-VL . For ERNIE-4.5 models, the technical report is here : https://yiyan.baidu.com/blog/publication/ERNIE_Technical_Report.pdf .(no arXiv link available)

vasqu commented on src/transformers/models/ernie4_5_vl/modular_ernie4_5_vl.py:
```diff
@@ -0,0 +1,1495 @@
+        self.num_experts = config.moe_num_experts
+
+        self.text_moe = Ernie4_5_VLSparseMoeBlock(config, intermediate_size=config.moe_intermediate_size[0])
+        self.vision_moe = Ernie4_5_VLSparseMoeBlock(config, intermediate_size=config.moe_intermediate_size[1])
```
I'd like to avoid passing the whole config to the language model tbh. Also, the vision model does not have anything vision-related - kinda have to mix things here or there but IMO the moe stuff should be kept together as it's specific to the language model.

vasqu commented on src/transformers/models/ernie4_5_vl/modular_ernie4_5_vl.py:
```diff
@@ -0,0 +1,1495 @@
+        head_dim = config.hidden_size // config.num_heads
+        self.rotary_pos_emb = Ernie4_5_VLVisionRotaryEmbedding(head_dim // 2)
+
+        self.ln = nn.LayerNorm(config.hidden_size, eps=config.vision_rms_norm_eps)
```
Do we want to? They just used that name and it's fairly standard - qwen didn't use that additional norm

vasqu commented on src/transformers/models/ernie4_5_vl/video_processing_ernie4_5_vl.py:
```diff
@@ -0,0 +1,469 @@
+                "Decrease `num_frames` or `fps` for sampling."
+            )
+
+        # same as `np.linspace(start=0, stop=total_num_frames, num=num_frames + 1).astype(int)[:-1]`
```
Oops, old notes lol

vasqu committed:
Pablo's review

vasqu committed:
style

vasqu committed:
fix device/dtype stuff that is no longer needed

vasqu committed:
revert vision property rm, necessary for composite sdpa test

molbap commented on src/transformers/models/ernie4_5_vl/convert_ernie4_5_vl_to_hf.py:
```diff
@@ -0,0 +1,460 @@
+    convert_config(args.checkpoint_path, args.pytorch_dump_folder_path)
+
+    # if args.convert_preprocessor:
+    #    convert_processor(args.checkpoint_path, args.pytorch_dump_folder_path)
```
just cleanup
```suggestion
    convert_config(args.checkpoint_path, args.pytorch_dump_folder_path)
```

molbap commented on src/transformers/models/ernie4_5_vl/video_processing_ernie4_5_vl.py:
```diff
@@ -0,0 +1,469 @@
+
+
+logger = logging.get_logger(__name__)
+DEFAULT_FONT_PATH = "AntonV/ernie4_5_fonts"
```
And besides `pix2struct` has the same structure (uses "arial.ttf", and font is saved in younes' own dir. Not a great pattern, let's break away from it

vasqu replied:
See https://github.com/huggingface/transformers/pull/39585/commits/36c1950307cbec97206ae69e2f24b9f05f2b4610

It got a bit more complicated than I initially thought but the current solution overrides `get_video_processor_dict` and adjusts an associated font on the fly based on the directory your processor is saved at. This requires users to save the font separately within the same repo but I think this is the best solution. Note that I needed to adjust the class for this in the base processor class so that it properly works (nothing crazy but just to keep in mind).

molbap commented on src/transformers/models/ernie4_5_vl/video_processing_ernie4_5_vl.py:
```diff
@@ -0,0 +1,469 @@
+            size["longest_edge"] = max_pixels
+            size.pop("max_pixels", None)
+        if "shortest_edge" not in size or "longest_edge" not in size:
+            raise ValueError("size must contain 'shortest_edge' and 'longest_edge' keys.")
```
Forever == until we can convert on load, hehe.

molbap commented on src/transformers/models/ernie4_5_vl/video_processing_ernie4_5_vl.py:
```diff
@@ -0,0 +1,461 @@
+        Can be overridden by subclasses to customize the processing of kwargs.
+        """
+        if size is not None and ("shortest_edge" not in size or "longest_edge" not in size):
+            raise ValueError("size must contain 'shortest_edge' and 'longest_edge' keys.")
```
do we need to raise both in the init and here?

molbap commented on src/transformers/models/ernie4_5_vl/video_processing_ernie4_5_vl.py:
```diff
@@ -0,0 +1,469 @@
+
+    def _render_image_with_timestamp(self, image: torch.Tensor, timestamp: str):
+        """Draws a black timestamp with a white border on the corner of the frame"""
+        image = render_text_on_image(to_pil_image(image), timestamp)
```
just measured out of curiosity, on GPU current order loses ~`6ms`/frame. Can add up, sure

vasqu replied:
We have at max `180 (frames) x 6ms = 1080ms = ~1,1s`. So yes it does add up but I would not prioritize this, wdyt about leaving a fixme/todo note there if anyone wants to fix/optimize this.

zucchini-nlp replied:
I think the fix is straightforward, if we move the `_render_image_with_timestamp` a few lines up. Unless I am missing smth. If it's indeed time-consuming, we can leave it for the next PR, I know this model is taking too much time to ship

molbap commented on utils/check_repo.py:
```diff
@@ -393,6 +394,7 @@
     "Qwen3OmniMoeTalkerModel",  # Building part of a bigger model
     "Qwen3OmniMoeThinkerForConditionalGeneration",  # Building part of a bigger model
     "Qwen3OmniMoeThinkerTextModel",  # Building part of a bigger model
+    "Ernie4_5_VLTextModel",  # Building part of a bigger model
```
We always end up adding stuff here - maybe we ought to update the checker one day? (out of scope here)

molbap reviewed (commented):
github has been a bit peculiar, been struggling with the new interface. Left very few comments, most of it is + what was already mentioned by Raushan is doable now that the weight loader https://github.com/huggingface/transformers/pull/41580 is merged. We can either merge now and use the converter later (by its nature, the order does not matter, we won't break BC), or merge later once the changes from converter are merged here, however you prefer

vasqu committed:
fixup few smaller things + refactor how we load the font entirely (based on font name with expected associated file at same repo)

vasqu commented on src/transformers/models/modernbert/modeling_modernbert.py:
Ignore this, local fa installation clashing with newest torch and modernbert is the only model which forces loading FA - should be rewritten at some point

vasqu commented on tests/models/ernie4_5_vl/test_processing_ernie4_5_vl.py:
```diff
@@ -0,0 +1,358 @@
+        ]
+
+        formatted_prompt = processor.apply_chat_template(messages, add_generation_prompt=True, tokenize=False)
+        self.assertEqual(len(formatted_prompt), 1)
```
Good question, honestly not sure anymore lol. Doesn't hurt to keep it either way

vasqu commented on src/transformers/models/ernie4_5_vl/modular_ernie4_5_vl.py:
```diff
@@ -0,0 +1,1495 @@
+                            batch_offset + (temporal_offset) * spatial_size,
+                            batch_offset + (temporal_offset + 1) * spatial_size,
+                        )
+                    )
```
That's for next week me, getting too late

vasqu committed:
remove bc min max pixels --> less modular on processor parts but way cleaner code

vasqu committed:
fix fps and add fixme to the inefficient conversion stuff

vasqu committed:
rope

vasqu committed:
style

vasqu committed:
copies and last rope stuff i fogot

vasqu committed:
revert glm4v copies

vasqu committed:
fix

vasqu committed:
simplify temporal slicing and add more descriptions

vasqu committed:
that ":" :cry:

vasqu committed:
Merge branch 'main' into ernie_vl

vasqu committed:
Merge branch 'main' into ernie_vl

vasqu committed:
fixup init

vasqu committed:
Merge branch 'main' into ernie_vl

vasqu committed:
conversion for moe split and merge + general renamings etc -- encountering OOM (automap maybe?)

vasqu committed:
wrong order whoops

vasqu committed:
style

vasqu committed:
copies

vasqu committed:
Merge branch 'main' into ernie_vl

vasqu committed:
fix init

vasqu committed:
fix

vasqu committed:
fix

vasqu committed:
allow the resolved path to be passed to explicit video processor classes and refactor how we load them for ernie

vasqu committed:
simplify

vasqu committed:
shoot, I need it there as well

vasqu committed:
better err handling

vasqu committed:
style

vasqu committed:
Merge branch 'main' into ernie_vl

vasqu committed:
Merge branch 'main' into ernie_vl

vasqu committed:
initial fixes after merge

vasqu committed:
working loading version

vasqu committed:
cleanup

vasqu committed:
change moe order and fix vl version

vasqu committed:
reverse op is mapping incorrectly TODO

vasqu committed:
reverse loading somewhat works, name conversion has issues it seems :eyes:

vasqu committed:
fix renaming issue, slow tests pass (except the integration ones ~ expected due to fused weights)

vasqu committed:
conversion mapping with native features + remove conversion mapping restriction

vasqu committed:
Merge branch 'main' into ernie_vl

vasqu committed:
add test for new conversion

vasqu committed:
style

vasqu committed:
update conversion

vasqu committed:
fix integration tests, remove fa tests

vasqu committed:
fix

vasqu committed:
update docs a bit

vasqu committed:
style

vasqu committed:
fix ernie moe and routing ernie series

vasqu committed:
style

zucchini-nlp commented on src/transformers/models/ernie4_5_vl/image_processing_ernie4_5_vl.py:
```diff
@@ -0,0 +1,436 @@
+                - `"none"` or `ChannelDimension.NONE`: image in (height, width) format.   - `"none"` or `ChannelDimension.NONE`: image in (height, width) format.
+        """
+        images = self.fetch_images(images)
+        images = make_list_of_images(images)
```
not needed anymore since we fetched and converted to list in `preprocess`?

vasqu committed:
fix rope warning

vasqu committed:
i fucked up again pain

vasqu committed:
update expectations

vasqu committed:
remove EP, broken atm be it sole or in combination with TP

vasqu committed:
update docs a bit

zucchini-nlp reviewed (commented):
huge work, the model has several special cases to handle. I went over the modular and processing code only, just a quick review with major comments

Would love to check out how on-the-fly conversion works as well, maybe tomorrow

vasqu committed:
first part of addressing review comments

vasqu committed:
fixup

vasqu committed:
fix vid processor

vasqu committed:
fix font saving

vasqu committed:
readd decorators oops

vasqu commented on src/transformers/models/ernie4_5_vl/configuration_ernie4_5_vl.py:
```diff
@@ -0,0 +1,303 @@
+    "Ernie4_5_VLConfig",
+    "Ernie4_5_VLTextConfig",
+    "Ernie4_5_VLVisionConfig",
+]
```
I can move them to modular but I cannot inherit them either way. They were originally very similar to QwenVL/GlmV but still have small diffs making them incompatible, e.g. Vision -> rms_norm, Text -> Moe namings differ slightly or the original Ernie Moe has a different handling of the moe intermediate size.

Not opposed to moving them but Idk, doesn't feel super worth it in this case.

zucchini-nlp replied:
if it's a few variables renamed/added/deleted, it can still be handled in modular imo. At least for vision I see a lot of overlapping with simple vit-like models. Prob the text overlaps with some MoE configs?

vasqu replied:
Moved to modular, the vision part is inheritable but the text part is hardly because only the `Glm Vision Moe` has the proper super call, but it barely matches. While Ernie Moe has the proper overlap, but not the super call. 

If we could do 2 super calls, i.e. one for inheriting the attributes and one for the final config super calls then it would be possible.

vasqu committed:
add mm token type id shortcut

vasqu commented on src/transformers/models/ernie4_5_vl/modular_ernie4_5_vl.py:
```diff
@@ -0,0 +1,1208 @@
+                        input_token_type.append("video")
+                    else:
+                        input_token_type.append("text")
+
```
Added in 0dcfe8b1ca51276fa820fadd6a898f2b2938b34a

It will speed up the first iteration which would be this awful loop. During decoding, we only have 1 token so it's not too bad IMO. I recoded the groups into their literals to work as tensors (0 == text, 1 == image, 2 == video). I haven't measured the speed gains but it's obvious that it's faster. NOTE: We only have this fast path if we use the processor (which should be the default case), otherwise we still have the slow path.

For this to properly work, I also added a workaround for the moe token types, see `get_moe_mm_token_type_ids`

vasqu committed:
always compose mm token type ids if needed

vasqu committed:
move config to modular

vasqu committed:
Merge branch 'main' into ernie_vl

vasqu committed:
fix loading by enforcing correct order

vasqu committed:
fix

vasqu committed:
Merge branch 'main' into ernie_vl

zucchini-nlp commented on docs/source/en/model_doc/ernie4_5_vl.md:
```diff
@@ -0,0 +1,221 @@
+- Dedicated Vision Experts
+- Shared Experts
+
+This architecture has the advantage to enhance multimodal understanding without compromising, and even improving, performance on text-related tasks. TODO image of the moe?
```
just so it's not forgotten, a todo here 😉

ArthurZucker commented on src/transformers/models/ernie4_5_vl/modular_ernie4_5_vl.py:
```diff
@@ -0,0 +1,1951 @@
+        position_ids_expanded = position_ids[:, :, None, :].float()  # shape (3, bs, 1, positions)
+
+        device_type = x.device.type if isinstance(x.device.type, str) and x.device.type != "mps" else "cpu"
+        with torch.autocast(device_type=device_type, enabled=False):  # Force float32
```
we added a maybe_autocast decorator!

ArthurZucker commented on src/transformers/models/ernie4_5_vl/modular_ernie4_5_vl.py:
```diff
@@ -0,0 +1,1951 @@
+    """
+
+
+class Ernie4_5_VLImageProcessor(Glm4vImageProcessor):
```
to remove we only need the fast one

ArthurZucker reviewed (approved):
Overall LGTM! Great work everyone!

zucchini-nlp reviewed (approved):
A couple more nits to be not forgotten before merging 😄

vasqu commented on src/transformers/models/ernie4_5_vl/modular_ernie4_5_vl.py:
```diff
@@ -0,0 +1,1951 @@
+
+        return mm_token_type_ids
+
+    def get_moe_mm_token_type_ids(self, original_mm_token_type_ids: Optional[torch.IntTensor]):
```
The issue is that I use `mm_token_type_ids` for
1. mRoPE positions to avoid looping over all tokens manually
2. Split Text / Vision MoE to differentiate which tokens go where

The first considers `image_token` + `video_token` as valid vision tokens. However, the moe considers additionally `image_start/end` + `video_start/end` as vision tokens. That method is the cleanest way IMO to do it ourselves without much effort. 

Should we introduce another kwarg here explicitly then, e.g. `moe_mm_token_type_ids`?

vasqu committed:
address first bunch of comments

vasqu committed:
Merge branch 'main' into ernie_vl

vasqu committed:
smaller comments

vasqu committed:
let's make moe layer types, ill fix modular in a second

vasqu committed:
modular

vasqu committed:
style

vasqu committed:
Merge branch 'main' into ernie_vl

vasqu committed:
renamed version along a few fixes in conversion and processor tests

vasqu committed:
fix

vasqu committed:
style + decorator

vasqu committed:
fix tokenizer handling of additional special tokens

vasqu committed:
style

vasqu committed:
fix doc refs

vasqu committed:
test fix

vasqu committed:
fix

vasqu committed:
was this too breaking?

vasqu committed:
fix conversion via workaround for now

vasqu committed:
Merge branch 'main' into ernie_vl

vasqu committed:
post merge fix

vasqu committed:
revert a few tok things (additional_special_tokens), updated conversion

vasqu committed:
Merge main into ernie_vl

ArthurZucker reviewed (commented):
Tokenization changes look good but would be better to put in another PR 🙏🏻

vasqu committed:
fix video processing loading logic

add exception for auto class (reload config as we have a circular dep on finding which class we have, i.e. we need to load to find the class then load with specific logic)

remove some original ideas

vasqu committed:
style

vasqu committed:
processor path change

vasqu committed:
Merge branch 'main' into ernie_vl

vasqu committed:
add small dummy integration tests

vasqu committed:
Merge branch 'main' into ernie_vl

vasqu committed:
style

vasqu committed:
fix rope modeling to follow qwen2 vl instead + change auto loading to specifically load via pretrained (overridable from pretrained for auto classes)

vasqu committed:
seems to be skipped in other similar vlms

vasqu committed:
Merge branch 'main' into ernie_vl

vasqu committed:
small conversion updates and adjust max vram usage during the big integration test

vasqu committed:
update test paths

vasqu committed:
style

vasqu committed:
style attmpt 2

vasqu committed:
docs

vasqu committed:
Merge branch 'main' into ernie_vl

vasqu committed:
trigger ci

vasqu commented:
run-slow: ernie4_5_vl_moe

molbap commented on src/transformers/models/ernie4_5_vl_moe/modular_ernie4_5_vl_moe.py:
```diff
@@ -0,0 +1,1893 @@
+
+        if isinstance(text_config, dict):
+            self.text_config = self.sub_configs["text_config"](**text_config)
+        elif isinstance(vision_config, Ernie4_5_VL_MoeTextConfig):
```
```suggestion
        elif isinstance(text_config, Ernie4_5_VL_MoeTextConfig):
```

molbap commented on src/transformers/models/ernie4_5_vl_moe/modular_ernie4_5_vl_moe.py:
```diff
@@ -0,0 +1,1893 @@
+            # We assume that all images have the same channel dimension format.
+            input_data_format = infer_channel_dimension_format(images[0])
+
+        height, width = get_image_size(images[0], channel_dim=input_data_format)
```
we also assume that all images have same base size, or should get_image_size be in the loop?

molbap reviewed (approved):
Read, re-read, re-re-read, I think this series is converging hehe. I left a few small comments, apologies if it's already treated above. 
massive work!

vasqu commented on src/transformers/models/ernie4_5_vl_moe/processing_ernie4_5_vl_moe.py:
Yea it's quite hacky especially since video and image treat with different final dimensions. The mm_token_type ids will have to be used at some point in all VLMs, makes it compatible with vLLM iirc

vasqu committed:
review

vasqu committed:
Merge branch 'main' into ernie_vl

vasqu committed:
post merge fixes

ArthurZucker commented on src/transformers/models/ernie4_5_vl_moe/video_processing_ernie4_5_vl_moe.py:
```diff
@@ -0,0 +1,594 @@
+                f"loading configuration file {video_processor_file} from cache at {resolved_video_processor_file}"
+            )
+
+        return video_processor_dict, kwargs
```
cc @yonigozlan this is something we want to be made easier: processor should define just the file extension / name and from pretrained downloads / finds everything for itt!

vasqu committed:
fix

vasqu committed:
safety

vasqu committed:
fix test

vasqu committed:
style

vasqu committed:
oops

vasqu committed:
fix

vasqu committed:
...

ArthurZucker reviewed (approved):
Huge work! A complicated model glad you got it to the finish line! 🤗

vasqu commented on src/transformers/models/ernie4_5_vl_moe/modular_ernie4_5_vl_moe.py:
```diff
@@ -0,0 +1,1905 @@
+        self.do_convert_rgb = do_convert_rgb
+
+    def _preprocess(
+        self,
```
I think only the fast processor covers these, the slow one does not. Hence the many kwargs, it's tech debt from the glmv4, qwen2+ vl series

vasqu committed:
simplify the config init for moe pattern

vasqu committed:
gonna be fixed by #42963

hiyouga commented on src/transformers/models/ernie4_5_vl_moe/modeling_ernie4_5_vl_moe.py:
```diff
@@ -0,0 +1,1893 @@
+        )
+
+        with maybe_autocast(device_type=device_type, enabled=False):  # Force float32
+            router_logits = F.linear(hidden_states.float(), self.weight)
```
@vasqu It works for me. Thank you for reference
